### ISSUE TITLE
How to get a Confusion Matrix

### ISSUE THREAD

**Issue by mmmichl** [PyTorchLightning/lightning-flash#899]
## ❓ Questions and Help
I made an image classifier with Lightning Flash and now I would like to see the confusion matrix of my classes. I am struggling to get this properly working.

I tried to add the metric to the `ImageClassifier` but as described in the documentation, only scalars as metrics are allowed - so this attempt failed.
```python
model = ImageClassifier(
    num_classes=datamodule.num_classes,
    metrics=torchmetrics.ConfusionMatrix(datamodule.num_classes, compute_on_step=False),
)
```

Is there a proper way to get a confusion matrix?

If not, I would highly appreciate if there would be way to get the raw data from `trainer.validate` and `trainer.test` so that one can easily calculate their own metric.
```python
# suggestion
val_res = trainer.validate(model, datamodule=datamodule, raw_results=True)
# >>> ({metric=99, ...}, ['class1, 'class2', ...], ['class2', 'class2', ...])
```

cheers and thanks for the library!

 - OS: [macOS, Linux]
 - Packaging [pip]
 - Version [e.g. 0.5.0]

**Comment by dmarx** [comment]
adding a minimal working example that reproduces the error I suspect Michl enocuntered:

```
# pip install lightning-flash[image] torchmetrics

import flash
import torch
from flash.image import ImageClassifier, ImageClassificationData
from torchvision.datasets import CIFAR10
import torchmetrics

datamodule = ImageClassificationData.from_datasets(
        train_dataset=CIFAR10('.', download=True, train=True),
    )

model = ImageClassifier(
    num_classes=10,
    metrics=torchmetrics.ConfusionMatrix(10, compute_on_step=False),
)

trainer = flash.Trainer(max_steps=30, gpus=torch.cuda.device_count())
trainer.finetune(model, datamodule=datamodule, strategy="freeze")
```

Produces the following error (traceback truncated):

```
...

/usr/local/lib/python3.7/dist-packages/pytorch_lightning/trainer/connectors/logger_connector/result.py in metrics(self, on_step)
    568             # populate progress_bar metrics. convert tensors to numbers
    569             if result_metric.meta.prog_bar:
--> 570                 metrics[MetricSource.PBAR][forked_name] = metrics_to_scalars(value)
    571 
    572         return metrics

/usr/local/lib/python3.7/dist-packages/pytorch_lightning/utilities/metrics.py in metrics_to_scalars(metrics)
     38         return value.item()
     39 
---> 40     return apply_to_collection(metrics, torch.Tensor, to_item)

/usr/local/lib/python3.7/dist-packages/pytorch_lightning/utilities/apply_func.py in apply_to_collection(data, dtype, function, wrong_dtype, include_none, *args, **kwargs)
     94     # Breaking condition
     95     if isinstance(data, dtype) and (wrong_dtype is None or not isinstance(data, wrong_dtype)):
---> 96         return function(data, *args, **kwargs)
     97 
     98     elem_type = type(data)

/usr/local/lib/python3.7/dist-packages/pytorch_lightning/utilities/metrics.py in to_item(value)
     34         if value.numel() != 1:
     35             raise MisconfigurationException(
---> 36                 f"The metric `{value}` does not contain a single element, thus it cannot be converted to a scalar."
     37             )
     38         return value.item()

MisconfigurationException: The metric `tensor([[6., 0., 1., 0., 3., 0., 0., 1., 3., 1.],
        [1., 0., 0., 1., 2., 0., 0., 5., 3., 2.],
        [1., 0., 0., 2., 2., 0., 0., 4., 0., 0.],
        [1., 0., 1., 0., 5., 0., 0., 1., 1., 0.],
        [2., 1., 2., 0., 2., 0., 0., 3., 3., 2.],
        [1., 0., 0., 1., 3., 1., 0., 1., 1., 2.],
        [1., 0., 0., 0., 3., 0., 0., 0., 2., 0.],
        [3., 0., 0., 0., 2., 0., 0., 3., 5., 2.],
        [6., 0., 0., 0., 1., 0., 0., 0., 1., 5.],
        [3., 0., 2., 0., 2., 0., 0., 1., 0., 6.]])` does not contain a single element, thus it cannot be converted to a scalar.
```

**Comment by dmarx** [comment]
@mmmichl Do you just need a single confusion matrix after model training has completed, or are you trying to calculate confusion matrices at regular intervals throughout the logging process? When you attach metrics to the model, flash assumes you are using those metrics to monitor the training process and will compute and log them at regular intervals. If you just want to compute the confusion matrix at the end of training, the easiest method would be to use the functional metric API after training has completed. I think the default lightning logger doesn't support logging confusion matrices (or any other non-scalar metrics, e.g. image tensors). If you want to compute and log the confusion matrix throughout training, we might be able to achieve that with a custom logger.

**Comment by MinWang1997** [comment]
Could you add the feature of default logger to log other metrics such as iou?


Thank you!

**Comment by mmmichl** [comment]
@dmarx I want to see the confusion matrix at the end of the training as validation. So there is no need to log the confusion matrix during the training for me.

Yeah, I did/tried something like that. But this is not very intuitive or too easy to accomplish, as I also use the `ImageClassificationData` data module. It took me couple of days to extract the validation data and to get the mapping of the labels:

```python
from flash.image import ImageClassifier, ImageClassificationData
import flash

datamodule = ImageClassificationData.from_data_frame(
    'filename',
    'class',
    df,
    data_dir,
    val_split=0.2
)

model = ImageClassifier(num_classes=datamodule.num_classes)
trainer = flash.Trainer(max_epochs=10)
trainer.finetune(model, datamodule=datamodule, strategy="freeze_unfreeze")

# now the ugly part
pred_res = trainer.predict(model, dataloaders=datamodule.val_dataloader)
preds = list(chain(*pred_res)) # merge the batches

from flash.core.data.data_source import LabelsState
actual_codes = datamodule.val_dataset
labels = datamodule.data_source.get_state(LabelsState).labels
actuals = [labels[i['target']] for i in actual_codes]

from sklearn.metrics import confusion_matrix
cm = confusion_matrix(actuals, preds, labels=labels)
```

This works, but compared to the training code, it is looks hacky and feels not right. But as I am new to this library I could miss a proper way to do things like that?!

**Comment by dmarx** [comment]
I think the "idiomatic lightning" approach here would be to use a callback invoked `on_train_end`: https://pytorch-lightning.readthedocs.io/en/latest/extensions/callbacks.html

**Comment by mmmichl** [comment]
We switched to pytorch lightning, where we handled the metrics on our own.
